clients/client-testextension/src/main/java/org/projectnessie/client/ext/NessieClientResolver.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (C) 2022 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.ext;
+
+import java.io.Serializable;
+import java.net.URI;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.projectnessie.client.api.NessieApiV1;
+import org.projectnessie.client.http.HttpClientBuilder;
+
+/**
+ * A base class for extensions that manage a Nessie test execution environment. This class injects
+ * Nessie java client instances and/or URIs based on the specific environment details provided by
+ * concrete subclasses.
+ *
+ * @see NessieUri
+ * @see NessieApiProvider
+ */
+public abstract class NessieClientResolver implements ParameterResolver {
+
+  protected abstract URI findBaseUri(ExtensionContext extensionContext);
+
+  private boolean isNessieUri(ParameterContext parameterContext) {
+    return parameterContext.isAnnotated(NessieUri.class);
+  }
+
+  private boolean isNessieClient(ParameterContext parameterContext) {
+    return parameterContext.getParameter().getType().isAssignableFrom(NessieApiProvider.class);
+  }
+
+  @Override
+  public boolean supportsParameter(
+      ParameterContext parameterContext, ExtensionContext extensionContext)
+      throws ParameterResolutionException {
+    return isNessieUri(parameterContext) || isNessieClient(parameterContext);
+  }
+
+  @Override
+  public Object resolveParameter(
+      ParameterContext parameterContext, ExtensionContext extensionContext)
+      throws ParameterResolutionException {
+    if (isNessieUri(parameterContext)) {
+      return findBaseUri(extensionContext);
+    }
+
+    if (isNessieClient(parameterContext)) {
+      return makeClient(extensionContext);
+    }
+
+    throw new IllegalStateException("Unsupported parameter: " + parameterContext);
+  }
+
+  private NessieApiProvider makeClient(ExtensionContext extensionContext) {
```

Comment:
nit: naming wise this also reads odd... makeClient returns something that can build an api instance :sweat_smile: 
maybe `NessieApiFactory` and `get` becomes `build(..)` ? "provider" sounds like you cant influence the returned object imo

Reply:
renaming...